libtenzir/include/tenzir/cast.hpp

```diff
@@ -335,6 +406,53 @@ struct cast_helper<record_type, record_type> {
     };
     return impl(impl, to_type, "");
   }
+
+  static auto cast_value(const record_type& from_type, const record& in,
+                         const record_type& to_type) noexcept
+    -> caf::expected<type_to_data_t<record_type>> {
+    if (from_type == to_type)
+      return in;
+    // NOLINTNEXTLINE
+    auto impl
+      = [&](const auto& impl, const record_type& to_type,
+            std::string_view key_prefix) noexcept -> caf::expected<record> {
+      auto ret = record{};
+      for (const auto& to_field : to_type.fields()) {
+        const auto key = key_prefix.empty()
+                           ? std::string{to_field.name}
+                           : fmt::format("{}.{}", key_prefix, to_field.name);
+        if (const auto* r = caf::get_if<record_type>(&to_field.type)) {
+          auto maybe_nested_record = impl(impl, *r, key);
+          if (not maybe_nested_record)
+            return std::move(maybe_nested_record.error());
+          ret[to_field.name] = std::move(*maybe_nested_record);
+          continue;
+        }
+        const auto index = from_type.resolve_key(key);
+        if (not index) {
+          // The field does not exist, so we insert a null.
+          ret[to_field.name] = caf::none;
+          continue;
+        }
+        // The field exists, so we can insert the cast column.
+        auto input_at_path = descend(std::addressof(in), key);
+        if (not input_at_path)
+          return std::move(input_at_path.error());
+        if (**input_at_path == caf::none) {
```

Comment:
Please add an assertion that it is safe to dereference twice here.

Reply:
done in 12dd25984f505a5630e02fbf12bb0785043bba53